modules/core/include/opencv2/core/quaternion.hpp

```diff
@@ -0,0 +1,1949 @@
+     * and \f$||q||\f$ represents the norm of \f$q\f$
+     *
+     * A unit quaternion is usually represents rotation, which has the form:
+     *      \f[q = [\cos\theta, u_x\sin\theta,u_y\sin\theta,  u_z\sin\theta].\f]
+     *
+     * To create a quaternion representing the rotation around the axis \f$\boldsymbol{v}\f$
+     * with angle \f$\psi\f$, you can use
+     * ```
+     * double angle = CV_PI;
+     * cv::Vec3d{0, 0, 1};
+     * cv::Quatd q(angle, axis);
+     * ```
+     *
+     * You can simply use four same type number to create a quaternion
+     * ```
+     * cv::Quatd q(1, 2, 3, 4);
+     * ```
+     * Or a Vec4d or Vec4f vec, then you can use
+     * ```
+     * cv::Quatd q(vec);
+     * ```
+     *
+     * If you already have a 3x3 rotate matrix R, then you can use
+     * ```
+     * cv::Quatd q(R);
+     * ```
+     *
+     * If you already have a Rodrigues vector rvec, then you can use
+     * ```
+     * cv::Quatd q(rvec);
+     * ```
+     *
+     * To extrace the rotate matrix from quaternion, see toRotMat3x3()
+     *
+     * To extrace the Vec4d or Vec4f, see toVec()
+     *
+     * To extrace the Rodrigues vector, see toRodrigues()
+     *
+     * If there are two quaternions \f$q_0, q_1\f$ interpolation is needed, you can use nlerp(), slerp() or spline()
+     * ```
+     * cv::nlerp(q0, q1, t)
+     *
+     * cv::slerp(q0, q1, t)
+     *
+     * cv::spline(q0, q0, q1, q1, t)
+     * ```
+     * spline can show the rotation smoothness between multiple quaternions
+     *
+     *
+     * Three ways to get a element in Quaternion
+     * ```
+     * Quatf q(1,2,3,4);
+     * std::cout << q.w << std::endl; // w=1,x=2,y=3,z=4
+     * std::cout << q[0] << std::endl; // q[0]=1, q[1]=2,q[2]=3,q[3]=4
+     * std::cout << q.at(0) << std::endl;
+     * ```
+     */
+    template <typename _Tp> class Quat;
+    template <typename S>
+    std::ostream& operator<<(std::ostream&, const Quat<S>&);
+
+    template <typename _Tp>
+    class Quat
+    {
+    using value_type = typename std::enable_if<
+        std::is_same<float, _Tp>::value||
+        std::is_same<double, _Tp>::value,
+        _Tp
+        >::type;
+    public:
+        Quat() = default;
+
+        /**
+         * @brief From Vec4d or Vec4f
+         */
+        explicit Quat(const cv::Vec<_Tp, 4> &coeff);
+
+        /**
+         * @brief from four numbers
+         */
+        Quat(_Tp w, _Tp x, _Tp y, _Tp z);
+
+        /**
+         * @brief from a angle, axis and qNorm with default 1, axis will be normalized in this constrctor. And
+         * it generates
+         * \f[q = [\cos\theta, u_x\sin\theta,u_y\sin\theta,  u_z\sin\theta].\f]
+         */
+        Quat(const _Tp angle, const cv::Vec<_Tp, 3> &axis);
+
+        /**
+         * @brief from a 3x3 rotate matrix
+         */
+        explicit Quat(const cv::Mat &R);
+
+        /**
+         * @brief from Rodrigues vector
+         */
+        explicit Quat(const Vec<_Tp, 3> &rvec);
```

Comment:
Need to provide an explanation that Rodrigues vector is axis-by-angle

Reply:
OK. Thanks for your suggestions